pipelines.md

```diff
@@ -0,0 +1,358 @@
+
+```
+sketch
+  |> for [0..n] as i
+    |> drawOneSectorOfGear(i) as x
+  |> extrude(...)
+
+// or
+
+sketch
+  |> for [0..n] as i |> {
+    x = sin(i)
+    y = cos(i)
+    drawOneSectorOfGear(x, y)
+  } as x
+  |> extrude(...)
+```
+
+In both cases, `x` has type [T] if `drawOneSectorOfGear` returns a `T`.
+
+Extension: Assumes we can use `export as` to export a variable from a block. Tagging within a for loop produces a somewhat special variable: it refers to the current iteration's tag within the loop and an array of tags outside the loop. E.g.,
+
+```
+sketch
+  |> for [0..n] as i |> {
+    drawOneSectorOfGear(i) export as x
+      |> foo(x) // x has type T
+  }
+  |> bar(x)     // x has type [T]
+  |> extrude(...)
+```
+
+### Alternative syntax
+
+Use `in` instead of `|>` in the for loop syntax. Examples:
+
+```
+x = [0..5] as i in i * 2
+
+sketch
+  |> for [0..n] as i in
+    drawOneSectorOfGear(i)
+  |> extrude(...)
+
+sketch
+  |> for [0..n] as i in {
+    drawOneSectorOfGear(i)
+      |> foo()
+  }
+  |> extrude(...)
+```
+
+I think this works better in the block case, but not so well in the case with a single expression.
+
+Or we could make the braced block mandatory rather than optional with a single expression:
+
+```
+x = [0..5] as i { i * 2 }
+
+sketch
+  |> for [0..n] as i {
+    drawOneSectorOfGear(i)
+  }
+  |> extrude(...)
+
+sketch
+  |> for [0..n] as i {
+    drawOneSectorOfGear(i)
+      |> foo()
+  }
+  |> extrude(...)
+```
+
+Or we could have a mandatory braced block *and* either `|>` or `in`.
+
+And/or, we could use `for i in [...]` rather than `for [...] as i`. This is closer to other PLs and reads better, however, it does not reuse the `as` keyword and so requires a new keyword.
+
+### Infix operators
+
+Infix operators (e.g., `+`, `-`) can be used in pipelines with the lhs of the pipeline operator being used as the lhs side of the infix operator. E.g., `1 |> + 1` is equivalent to `1 + 1`. This is not really necessary by itself, but might be if in the future we support infix operators for 2D or 3D geometry, e.g., `sketch |> extrude(...) |> union something |> union somethingElse()`, assuming `a union b` is allowed.
+
+I expect this part of the proposal should be low priority to implement. I'm just including it here for completeness.
+
+## Alternative: rely less on pipelines
+
+[Example](https://github.com/KittyCAD/modeling-app/issues/2728#issuecomment-2361355398):
+
+```
+squareSketch = sketchOn('XY', [
+  line(4, 0), 
+  line(0, 4), 
+  line(-4, 0), 
+  line(0, -4),
+])
+```
+
+Advantages:
+
+* Simpler
+  - More declarative since we're declaring the result in one go rather than the steps to building it
+  - Less visual noise (no `|>`, etc.)
+  - Possibly fewer required steps
+* More expressive - user can manipulate arrays rather than just build up objects
+* Perhaps closer to the execution model of building a geometry model and sending that to the engine, rather than sending each step to the engine
```

Comment:
I think it'd also be simpler to optimize. E.g. one problem we're having right now is that we're considering moving some math calculations out of KCL's executor and into either engine API calls or engine WASM artifacts run in the same process but in a WASM runtime. If we could do all the line calculations in the same AST node we could reason better about which expensive calculations we need and which can be omitted.

Reply:
That's a good point!